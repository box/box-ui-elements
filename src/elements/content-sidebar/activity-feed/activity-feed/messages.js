/**
 * @flow
 * @file i18n messages
 * @author Box
 */

import { defineMessages } from 'react-intl';

const messages = defineMessages({
    annotationMissingError: {
        id: 'be.activitySidebar.activityFeed.annotationMissingError',
        defaultMessage: 'This comment no longer exists',
        description: 'Text to show when an annotation activity no longer exists',
    },
    feedInlineErrorTitle: {
        id: 'be.activitySidebar.activityFeed.feedInlineErrorTitle',
        defaultMessage: 'Error',
        description: 'Error title',
    },
    taskMissingError: {
        id: 'be.activitySidebar.activityFeed.taskMissingError',
        defaultMessage: 'This task no longer exists',
        description: 'Text to show when a task no longer exists',
    },
    commentMissingError: {
        id: 'be.activitySidebar.activityFeed.commentMissingError',
        defaultMessage: 'This comment no longer exists',
        description: 'Text to show when comment no longer exists',
    },
});

export default messages;
