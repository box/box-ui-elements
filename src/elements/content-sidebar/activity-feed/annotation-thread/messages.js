import { defineMessages } from 'react-intl';

const messages = defineMessages({
    errorFetchAnnotation: {
        id: 'be.annotattionThread.errorFetchAnnotation',
        description: 'Error message when an annotation fetch fails',
        defaultMessage: 'The annotation could not be fetched.',
    },
    errorEditAnnotation: {
        id: 'be.annotationThread.errorEditAnnotation',
        description: 'Error message when an annotation update fails',
        defaultMessage: 'This annotation could not be modified.',
    },
    errorDeleteAnnotation: {
        id: 'be.annotationThread.errorDeleteAnnotation',
        description: 'Error message when an annotation deletion fails',
        defaultMessage: 'There was an error deleting this item.',
    },
});

export default messages;
