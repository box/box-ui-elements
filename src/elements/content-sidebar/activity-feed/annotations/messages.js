import { defineMessages } from 'react-intl';

const messages = defineMessages({
    annotationActivityDeleteMenuItem: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityDeleteMenuItem',
        defaultMessage: 'Delete',
        description: 'Text to show on menu item to delete the annotation comment',
    },
    annotationActivityDeletePrompt: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityDeletePrompt',
        defaultMessage: 'Are you sure you want to permanently delete this comment?',
        description: 'Confirmation prompt text to delete the annotation comment',
    },
    annotationActivityEditMenuItem: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityEditMenuItem',
        defaultMessage: 'Modify',
        description: 'Text to show on menu item to edit the annotation comment',
    },
    annotationActivityPostedFullDateTime: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityPostedFullDateTime',
        defaultMessage: '{time, date, full} at {time, time, short}',
        description: 'Full data and time for annotation title',
    },
    annotationActivityPageItem: {
        id: 'be.contentSidebar.activityFeed.annotationActivityPageItem',
        defaultMessage: 'Page {number}',
        description: 'Annotation activity item link shown on annotation activity',
    },
    annotationActivityVersionLink: {
        id: 'be.contentSidebar.activityFeed.annotationActivityVersionLink',
        defaultMessage: 'Version {number}',
        description: 'Annotation activity item link shown on annotation activity for previous file version',
    },
    annotationActivityVersionUnavailable: {
        id: 'be.contentSidebar.activityFeed.annotationActivityVersionUnavailable',
        defaultMessage: 'Version Unavailable',
        description:
            'Annotation activity item link shown on annotation activity for previous file version that is unavailable',
    },
});

export default messages;
