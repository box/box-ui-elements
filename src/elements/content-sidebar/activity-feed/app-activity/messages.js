/**
 * @flow
 * @file i18n messages
 * @author Box
 */

import { defineMessages } from 'react-intl';

const messages = defineMessages({
    appActivityDeleteMenuItem: {
        id: 'be.appActivity.deleteMenuItem',
        defaultMessage: 'Delete',
        description: 'Text to display in the delete app activity submit button',
    },
    appActivityDeletePrompt: {
        id: 'be.appActivity.deletePrompt',
        defaultMessage: 'Are you sure you want to permanently delete this app activity?',
        description: 'Confirmation prompt text to delete app activity',
    },
    appActivityAltIcon: {
        id: 'be.appActivity.altIcon',
        defaultMessage: '{appActivityName} Icon',
        description: 'Alt message if app activity icon is missing or cannot load',
    },
});

export default messages;
