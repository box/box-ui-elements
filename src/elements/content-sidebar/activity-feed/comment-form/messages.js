/**
 * @flow
 * @file i18n messages
 * @author Box
 */

import { defineMessages } from 'react-intl';

const messages = defineMessages({
    commentCancel: {
        id: 'be.contentSidebar.activityFeed.commentForm.commentCancel',
        defaultMessage: 'Cancel',
        description: 'Text for cancel button',
    },
    commentLabel: {
        id: 'be.contentSidebar.activityFeed.commentForm.commentLabel',
        defaultMessage: 'Write a comment',
        description: 'Accessible label for comment input field',
    },
    commentPost: {
        id: 'be.contentSidebar.activityFeed.commentForm.commentPost',
        defaultMessage: 'Post',
        description: 'Text for post button',
    },
    commentWrite: {
        id: 'be.contentSidebar.activityFeed.commentForm.commentWrite',
        defaultMessage: 'Write a comment',
        description: 'Placeholder for comment input',
    },
    approvalAddAssignee: {
        id: 'be.contentSidebar.activityFeed.commentForm.approvalAddAssignee',
        defaultMessage: 'Add an assignee',
        description: 'Placeholder for approvers input',
    },
    approvalAddTask: {
        id: 'be.contentSidebar.activityFeed.commentForm.approvalAddTask',
        defaultMessage: 'Add Task',
        description: 'Label for checkbox to add approvers to a comment',
    },
    approvalAddTaskTooltip: {
        id: 'be.contentSidebar.activityFeed.commentForm.approvalAddTaskTooltip',
        defaultMessage:
            'Assigning a task to someone will send them a notification with the message in the comment box and allow them to approve or deny.',
        description: 'Tooltip text for checkbox to add approvers to a comment',
    },
    approvalAssignees: {
        id: 'be.contentSidebar.activityFeed.commentForm.approvalAssignees',
        defaultMessage: 'Assignees',
        description: 'Title for assignees input',
    },
    approvalDueDate: {
        id: 'be.contentSidebar.activityFeed.commentForm.approvalDueDate',
        defaultMessage: 'Due Date',
        description: 'Title for approvers due date input',
    },
    approvalSelectDate: {
        id: 'be.contentSidebar.activityFeed.commentForm.approvalSelectDate',
        defaultMessage: 'Select a date',
        description: 'Placeholder for due date input',
    },
    atMentionTip: {
        id: 'be.contentSidebar.activityFeed.commentForm.atMentionTip',
        defaultMessage: '@mention users to notify them.',
        description: 'Mentioning call to action displayed below the comment input',
    },
    atMentionTipDescription: {
        id: 'be.contentSidebar.activityFeed.commentForm.atMentionTipDescription',
        defaultMessage:
            'Use the @ symbol to mention users and use the up and down arrow keys to scroll through autocomplete suggestions.',
        description: 'Mentioning call to action detailed description for screen reader users',
    },
});

export default messages;
