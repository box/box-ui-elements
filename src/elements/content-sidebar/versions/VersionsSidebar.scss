@import '../../../styles/variables';

.bcs-Versions {
    .bcs-scroll-content {
        width: 100%;
    }

    .bcs-section-content:first-child {
        padding-top: 0;
    }

    .bdl-BackButton {
        margin-right: 5px;
    }

    .crawler {
        display: flex;
        align-items: center;
        height: 50px;
    }
}

.bcs-Versions-content {
    height: 100%;
    padding-right: 25px;
    padding-left: 25px;
}

.bcs-Versions-empty,
.bcs-Versions-maxEntries {
    padding-top: 10px;
    text-align: center;
}

.bcs-Versions-menu {
    .be & {
        padding-bottom: 20px;
    }
}

.btn-plain.bcs-Versions-upgrade {
    color: $bdl-box-blue;
    font-weight: bold;
}
