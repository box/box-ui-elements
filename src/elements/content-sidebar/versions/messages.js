// @flow strict

import { defineMessages } from 'react-intl';

const messages = defineMessages({
    fileRequestDisplayName: {
        defaultMessage: 'File Request',
        description:
            'name of the File Request feature used to translate when a File Request is uploaded by the service',
        id: 'be.sidebarVersions.fileRequestDisplayName',
    },
    versionsEmpty: {
        id: 'be.sidebarVersions.empty',
        description: 'Message to display when no versions are available',
        defaultMessage: 'No prior versions are available for this file.',
    },
    versionsPriorWeek: {
        id: 'be.sidebarVersions.priorWeek',
        description: 'Header to display for group of versions created in the prior week',
        defaultMessage: 'Last Week',
    },
    versionsThisMonth: {
        id: 'be.sidebarVersions.thisMonth',
        description: 'Header to display for group of versions created in the current month',
        defaultMessage: 'This Month',
    },
    versionsToday: {
        id: 'be.sidebarVersions.today',
        description: 'Header to display for group of versions created yesterday',
        defaultMessage: 'Today',
    },
    versionsYesterday: {
        id: 'be.sidebarVersions.yesterday',
        description: 'Header to display for group of versions created today',
        defaultMessage: 'Yesterday',
    },
    versionsTitle: {
        id: 'be.sidebarVersions.title',
        description: 'Title for the preview versions sidebar',
        defaultMessage: 'Version History',
    },
    versionCurrent: {
        id: 'be.sidebarVersions.current',
        defaultMessage: 'Current Version',
        description: 'Label for the current version item in the version history list.',
    },
    versionActionDelete: {
        id: 'be.sidebarVersions.delete',
        defaultMessage: 'Delete',
        description: 'Label for the version delete action.',
    },
    versionActionDeleteError: {
        id: 'be.sidebarVersions.deleteError',
        defaultMessage: 'File version could not be deleted.',
        description: 'Error message for the version delete action.',
    },
    versionActionDisabledRetention: {
        id: 'be.sidebarVersions.disabledByRetention',
        defaultMessage: 'Disabled by retention policy',
        description: 'Tooltip message for actions disabled by retention policy.',
    },
    versionActionDownload: {
        id: 'be.sidebarVersions.download',
        defaultMessage: 'Download',
        description: 'Label for the version download action.',
    },
    versionActionDownloadError: {
        id: 'be.sidebarVersions.downloadError',
        defaultMessage: 'File version could not be downloaded.',
        description: 'Error message for the version download action.',
    },
    versionActionPreview: {
        id: 'be.sidebarVersions.preview',
        defaultMessage: 'Preview',
        description: 'Label for the version preview action.',
    },
    versionActionPromote: {
        id: 'be.sidebarVersions.promote',
        defaultMessage: 'Make Current',
        description: 'Label for the version promote action.',
    },
    versionActionPromoteError: {
        id: 'be.sidebarVersions.promoteError',
        defaultMessage: 'File version could not be made current.',
        description: 'Error message for the version promote action.',
    },
    versionActionRestore: {
        id: 'be.sidebarVersions.restore',
        defaultMessage: 'Restore',
        description: 'Label for the version restore action.',
    },
    versionActionRestoreError: {
        id: 'be.sidebarVersions.restoreError',
        defaultMessage: 'File version could not be restored.',
        description: 'Error message for the version restored action.',
    },
    versionActionToggle: {
        id: 'be.sidebarVersions.toggle',
        defaultMessage: 'Toggle Actions Menu',
        description: 'Label for the version actions dropdown menu toggle button.',
    },
    versionDeletedBy: {
        id: 'be.sidebarVersions.deletedBy',
        defaultMessage: 'Deleted by {name}',
        description: 'Message displayed for a deleted version. {name} is the user who performed the action.',
    },
    versionFetchError: {
        id: 'be.sidebarVersions.fetchError',
        defaultMessage: 'File versions could not be retrieved.',
        description: 'Error message for the versions fetch call.',
    },
    versionNotAvailable: {
        id: 'be.sidebarVersions.notAvailable',
        defaultMessage:
            'Sorry, version history is not available with your current account plan. To access versioning, select from one of our paid plans.',
        description:
            'Error message with upsell for the versions fetch call for users who dont have the version feature.',
    },
    versionLimitExceeded: {
        id: 'be.sidebarVersions.versionLimitExceeded',
        defaultMessage:
            'You are limited to the last {versionLimit, number} {versionLimit, plural, one {version} other {versions}}.',
        description: "Text displayed if a version exceeds the user's maximum allowed version count",
    },
    versionNumberBadge: {
        id: 'be.sidebarVersions.versionNumberBadge',
        defaultMessage: 'V{versionNumber}',
        description: 'Text to display in the version badge.',
    },
    versionNumberLabel: {
        id: 'be.sidebarVersions.versionNumberLabel',
        defaultMessage: 'Version number {versionNumber}',
        description: 'Label given to the version badge for screen readers.',
    },
    versionPromotedBy: {
        id: 'be.sidebarVersions.promotedBy',
        defaultMessage: 'Promoted from v{versionPromoted} by {name}',
        description: 'Message displayed for a restored version. {name} is the user who performed the action.',
    },
    versionRestoredBy: {
        id: 'be.sidebarVersions.restoredBy',
        defaultMessage: 'Restored by {name}',
        description: 'Message displayed for a restored version. {name} is the user who performed the action.',
    },
    versionServerError: {
        id: 'be.sidebarVersions.serverError',
        defaultMessage: 'Server Error',
        description: 'Heading for inline error messages displayed due to a remote server issue.',
    },
    versionAccessError: {
        id: 'be.sidebarVersions.accessError',
        defaultMessage: "You don't have access to versioning",
        description: 'Heading for inline error messages with upsell.',
    },
    versionUpgradeButton: {
        id: 'be.sidebarVersions.versionUpgradeButton',
        defaultMessage: 'Upgrade Now',
        description: 'Text telling user to upgrade now.',
    },
    versionUploadedBy: {
        id: 'be.sidebarVersions.uploadedBy',
        defaultMessage: 'Uploaded by {name}',
        description: 'Message displayed for an uploaded version. {name} is the user who performed the action.',
    },
    versionUserUnknown: {
        id: 'be.sidebarVersions.versionUserUnknown',
        defaultMessage: 'Unknown',
        description: 'Name displayed for unknown or deleted users.',
    },
    versionMaxEntries: {
        id: 'be.sidebarVersions.versionMaxEntries',
        defaultMessage: 'Version history is limited to the last {maxVersions} entries.',
        description: 'Max supported entries for version history',
    },
    versionRetentionDelete: {
        id: 'be.sidebarVersions.versionRetentionDelete',
        defaultMessage: 'Will be deleted {time} by retention policy.',
        description: 'Message describing when the version will be deleted due to an applied retention policy.',
    },
    versionRetentionIndefinite: {
        id: 'be.sidebarVersions.versionRetentionIndefinite',
        defaultMessage: 'Retained indefinitely by retention policy.',
        description: 'Message describing that the version retention policy is indefinite and will not expire.',
    },
    versionRetentionRemove: {
        id: 'be.sidebarVersions.versionRetentionRemove',
        defaultMessage: 'Retention policy expires on {time}.',
        description: 'Message describing when the version retention policy will expire.',
    },
});

export default messages;
