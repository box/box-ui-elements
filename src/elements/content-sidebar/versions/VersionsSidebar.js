/**
 * @flow
 * @file Versions Sidebar component
 * @author Box
 */

import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import type { MessageDescriptor } from 'react-intl';
import InlineError from '../../../components/inline-error';
import messages from './messages';
import SidebarContent from '../SidebarContent';
import VersionsMenu from './VersionsMenu';
import { BackButton } from '../../common/nav-button';
import PlainButton from '../../../components/plain-button/PlainButton';
import { DEFAULT_FETCH_END } from '../../../constants';
import { LoadingIndicatorWrapper } from '../../../components/loading-indicator';
import type { BoxItemVersion } from '../../../common/types/core';
import './VersionsSidebar.scss';

const MAX_VERSIONS = DEFAULT_FETCH_END;

type Props = {
    error?: MessageDescriptor,
    errorTitle?: MessageDescriptor,
    fileId: string,
    isLoading: boolean,
    onUpgradeClick?: () => void,
    parentName: string,
    versionCount: number,
    versionLimit: number,
    versions: Array<BoxItemVersion>,
};

const VersionsSidebar = ({ error, errorTitle, onUpgradeClick, isLoading, parentName, versions, ...rest }: Props) => {
    const showLimit = versions.length >= MAX_VERSIONS;
    const showVersions = !!versions.length;
    const showEmpty = !isLoading && !showVersions && !onUpgradeClick;
    const showError = !!error;

    return (
        <SidebarContent
            className="bcs-Versions"
            data-resin-component="preview"
            data-resin-feature="versions"
            title={
                <>
                    <BackButton data-resin-target="back" to={`/${parentName}`} />
                    <FormattedMessage {...messages.versionsTitle} />
                </>
            }
        >
            <LoadingIndicatorWrapper className="bcs-Versions-content" crawlerPosition="top" isLoading={isLoading}>
                {showError && (
                    <InlineError title={<FormattedMessage {...errorTitle} />}>
                        <FormattedMessage {...error} />
                        {onUpgradeClick && (
                            <PlainButton
                                onClick={onUpgradeClick}
                                data-resin-target="versioning_error_error_message_upgrade_cta"
                                type="button"
                                className="bcs-Versions-upgrade"
                            >
                                <FormattedMessage {...messages.versionUpgradeButton} />
                            </PlainButton>
                        )}
                    </InlineError>
                )}

                {showEmpty && (
                    <div className="bcs-Versions-empty">
                        <FormattedMessage {...messages.versionsEmpty} />
                    </div>
                )}

                {showVersions && (
                    <div className="bcs-Versions-menu">
                        <VersionsMenu versions={versions} {...rest} />
                    </div>
                )}
                {showLimit && (
                    <div className="bcs-Versions-maxEntries" data-testid="max-versions">
                        <FormattedMessage
                            {...messages.versionMaxEntries}
                            values={{
                                maxVersions: MAX_VERSIONS,
                            }}
                        />
                    </div>
                )}
            </LoadingIndicatorWrapper>
        </SidebarContent>
    );
};

export default VersionsSidebar;
