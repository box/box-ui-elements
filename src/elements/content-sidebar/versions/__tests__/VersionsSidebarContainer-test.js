import * as React from 'react';
import { shallow } from 'enzyme/build';
import messages from '../messages';
import openUrlInsideIframe from '../../../../utils/iframe';
import VersionsSidebar from '../VersionsSidebarContainer';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    withRouter: Component => Component,
}));

jest.mock('../../../common/api-context', () => ({
    withAPIContext: Component => Component,
}));

jest.mock('../../../../utils/iframe', () => ({
    __esModule: true,
    default: jest.fn(),
}));

const versions = [{ id: '123', name: 'Version 1' }, { id: '456', name: 'Version 2' }, { id: '789', name: 'Version 3' }];

describe('elements/content-sidebar/versions/VersionsSidebarContainer', () => {
    const versionsAPI = {
        addPermissions: jest.fn(),
        sortVersions: jest.fn(),
    };
    const api = {
        getVersionsAPI: () => versionsAPI,
    };
    const getWrapper = ({ ...props } = {}) => shallow(<VersionsSidebar api={api} fileId="12345" {...props} />);

    describe('componentDidUpdate', () => {
        let onVersionChange;
        let wrapper;
        let instance;

        beforeEach(() => {
            onVersionChange = jest.fn();
            wrapper = getWrapper({ onVersionChange, refreshIdentity: false });
            instance = wrapper.instance();
        });

        test('should verify the selected version id exists when it changes', () => {
            const version = { id: '45678' };
            const currentVersion = { id: '54321' };

            instance.verifyVersion = jest.fn();

            wrapper.setState({ versions: [currentVersion, version] });
            wrapper.setProps({ versionId: '45678' });

            expect(instance.verifyVersion).toHaveBeenCalled();
        });
    });

    describe('componentWillUnmount', () => {
        test('should forward verison id reset to the parent component', () => {
            const onVersionChange = jest.fn();
            const wrapper = getWrapper({ onVersionChange });

            wrapper.instance().componentWillUnmount();

            expect(onVersionChange).toBeCalledWith(null);
        });
    });

    describe('handleActionDelete', () => {
        test('should call api endpoint helpers', () => {
            const handleDelete = jest.fn();
            const wrapper = getWrapper({ onVersionDelete: handleDelete, versionId: '123' });
            const instance = wrapper.instance();
            const version = { id: '456' };

            instance.api.deleteVersion = jest.fn().mockResolvedValueOnce();
            instance.api.fetchData = jest.fn().mockResolvedValueOnce();
            instance.findVersion = jest.fn(() => version);
            instance.handleFetchSuccess = jest.fn();
            instance.handleDeleteSuccess = jest.fn();

            instance.handleActionDelete(version.id).then(() => {
                expect(instance.api.deleteVersion).toHaveBeenCalledWith(version);
                expect(instance.api.fetchData).toHaveBeenCalled();
                expect(instance.handleFetchSuccess).toHaveBeenCalled();
                expect(instance.handleDeleteSuccess).toHaveBeenCalledWith(version.id);
                expect(handleDelete).toHaveBeenCalledWith(version.id);
            });
        });
    });

    describe('handleActionDownload', () => {
        test('should call api endpoint helpers', () => {
            const downloadUrl = 'https://box.com/url';
            const handleDownload = jest.fn();
            const wrapper = getWrapper({ onVersionDownload: handleDownload, versionId: '123' });
            const instance = wrapper.instance();
            const version = { id: '456' };

            instance.api.fetchDownloadUrl = jest.fn().mockResolvedValueOnce(downloadUrl);
            instance.findVersion = jest.fn(() => version);

            instance.handleActionDownload(version.id).then(() => {
                expect(instance.api.fetchDownloadUrl).toHaveBeenCalledWith(version);
                expect(openUrlInsideIframe).toHaveBeenCalledWith(downloadUrl);
                expect(handleDownload).toHaveBeenCalledWith(version.id);
            });
        });
    });

    describe('handleActionPromote', () => {
        test('should call api endpoint helpers', () => {
            const handlePromote = jest.fn();
            const wrapper = getWrapper({ onVersionPromote: handlePromote, versionId: '123' });
            const instance = wrapper.instance();
            const version = { id: '456' };

            instance.api.fetchData = jest.fn().mockResolvedValueOnce();
            instance.api.promoteVersion = jest.fn().mockResolvedValueOnce();
            instance.findVersion = jest.fn(() => version);
            instance.handleFetchSuccess = jest.fn();
            instance.handlePromoteSuccess = jest.fn();

            instance.handleActionPromote(version.id).then(() => {
                expect(instance.api.promoteVersion).toHaveBeenCalledWith(version);
                expect(instance.api.fetchData).toHaveBeenCalled();
                expect(instance.handleFetchSuccess).toHaveBeenCalled();
                expect(instance.handlePromoteSuccess).toHaveBeenCalled();
                expect(handlePromote).toHaveBeenCalledWith(version.id);
            });
        });
    });

    describe('handleActionRestore', () => {
        test('should call api endpoint helpers', () => {
            const handleRestore = jest.fn();
            const wrapper = getWrapper({ onVersionRestore: handleRestore, versionId: '123' });
            const instance = wrapper.instance();
            const version = { id: '456' };

            instance.api.fetchData = jest.fn().mockResolvedValueOnce();
            instance.api.restoreVersion = jest.fn().mockResolvedValueOnce();
            instance.findVersion = jest.fn(() => version);
            instance.handleFetchSuccess = jest.fn();

            instance.handleActionRestore(version.id).then(() => {
                expect(instance.api.restoreVersion).toHaveBeenCalledWith(version);
                expect(instance.api.fetchData).toHaveBeenCalled();
                expect(instance.handleFetchSuccess).toHaveBeenCalled();
                expect(handleRestore).toHaveBeenCalledWith(version.id);
            });
        });
    });

    describe('handleFetchError', () => {
        test('should set state to default values with error message', () => {
            const wrapper = getWrapper();

            wrapper.instance().handleFetchError();

            expect(wrapper.state()).toEqual({
                error: messages.versionFetchError,
                isLoading: false,
                isWatermarked: false,
                versionCount: 0,
                versionLimit: Infinity,
                versions: [],
            });
        });
    });

    describe('handleFetchSuccess', () => {
        let file;
        let version;
        let currentVersion;
        let versionsWithCurrent;

        beforeEach(() => {
            file = { id: '12345', permissions: {}, version_limit: 10 };
            version = { id: '123', permissions: {} };
            currentVersion = { entries: [{ id: '321', permissions: {} }], total_count: 1 };
            versionsWithCurrent = { entries: [version, ...currentVersion.entries], total_count: 2 };

            versionsAPI.addPermissions.mockReturnValueOnce(versionsWithCurrent);
        });

        test('should set state with the updated versions', () => {
            const wrapper = getWrapper();
            const instance = wrapper.instance();

            instance.sortVersions = jest.fn().mockReturnValue(versionsWithCurrent.entries);
            instance.verifyVersion = jest.fn();
            instance.handleFetchSuccess([file, versionsWithCurrent]);

            expect(versionsAPI.addPermissions).toBeCalledWith(versionsWithCurrent, file);
            expect(instance.verifyVersion).toBeCalled();
            expect(instance.sortVersions).toBeCalledWith(versionsWithCurrent.entries);
            expect(wrapper.state()).toMatchObject({
                error: undefined,
                isLoading: false,
                isWatermarked: false,
                versionCount: 2,
                versionLimit: 10,
            });
            expect(wrapper.state('versions')).toBe(versionsWithCurrent.entries);
        });

        test('should set state with isWatermarked if file is watermarked', () => {
            const wrapper = getWrapper();
            const instance = wrapper.instance();
            const testFile = { ...file, watermark_info: { is_watermarked: true } };

            instance.verifyVersion = jest.fn();
            instance.handleFetchSuccess([testFile, versionsWithCurrent]);

            expect(wrapper.state('isWatermarked')).toBe(true);
        });
    });

    describe('findVersion', () => {
        test('should return the version stored in state if available', () => {
            const wrapper = getWrapper();
            const instance = wrapper.instance();

            wrapper.setState({ versions });

            expect(instance.findVersion('456')).toEqual(versions[1]);
            expect(instance.findVersion('abc')).toEqual(undefined);
        });
    });

    describe('refresh', () => {
        test('should refetch data when refresh is called', () => {
            const instance = getWrapper().instance();
            const fetchData = jest.fn();
            instance.fetchData = fetchData;

            instance.refresh();

            expect(fetchData).toHaveBeenCalled();
        });
    });

    describe('sortVersions', () => {
        test('should sort versions by their created date', () => {
            const instance = getWrapper().instance();
            const unsortedVersions = [
                { id: '123', created_at: '2018-10-02T13:01:24-07:00' },
                { id: '456', created_at: '2018-10-02T13:01:41-07:00' },
                { id: '789', created_at: '2018-11-29T17:47:57-08:00' },
            ];
            const sortedVersions = instance.sortVersions(unsortedVersions);

            expect(unsortedVersions).not.toBe(sortedVersions); // Sort call should create a new array
            expect(sortedVersions).toEqual([unsortedVersions[2], unsortedVersions[1], unsortedVersions[0]]);
        });
    });

    describe('verifyVersion', () => {
        const onVersionChange = jest.fn();

        test('should emit an onVersionChange event if the passed version is available', () => {
            const wrapper = getWrapper({ onVersionChange, versionId: '456' });

            wrapper.setState({ versions });
            wrapper.instance().verifyVersion();

            expect(onVersionChange).toHaveBeenCalledWith(versions[1], {
                currentVersionId: versions[0].id,
                updateVersionToCurrent: wrapper.instance().updateVersionToCurrent,
            });
        });

        test('should reset the selected version if the passed version is not available', () => {
            const wrapper = getWrapper({ onVersionChange, versionId: '99999' });
            const instance = wrapper.instance();

            instance.updateVersionToCurrent = jest.fn();

            wrapper.setState({ versions });
            instance.verifyVersion();

            expect(onVersionChange).not.toHaveBeenCalled();
            expect(instance.updateVersionToCurrent).toHaveBeenCalled();
        });
    });

    describe('render', () => {
        test('should match its snapshot', () => {
            const wrapper = getWrapper({ parentName: 'activity' });
            expect(wrapper).toMatchSnapshot();
        });
    });
});
