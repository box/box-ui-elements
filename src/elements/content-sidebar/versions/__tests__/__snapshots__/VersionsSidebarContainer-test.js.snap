// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/content-sidebar/versions/VersionsSidebarContainer render should match its snapshot 1`] = `
<VersionsSidebar
  isLoading={true}
  onDelete={[Function]}
  onDownload={[Function]}
  onPreview={[Function]}
  onPromote={[Function]}
  onRestore={[Function]}
  parentName="activity"
  versions={Array []}
/>
`;
